packages/core/elasticsearch/core-elasticsearch-client-server-internal/src/agent_manager.ts

```diff
@@ -0,0 +1,188 @@
+ * exposes methods that can modify the underlying pools, effectively impacting the connections of other Clients.
+ * @internal
+ **/
+export class AgentManager {
+  /**
+   * Store Agent instances by type and protocol, e.g.:
+   *
+   * data: {
+   *   'http:': [agentInstance1, agentInstance2],
+   *   'https:': [agentInstance3, agentInstance4]
+   * },
+   * monitoring: {
+   *   'http:': [],
+   *   'https:': [agentInstance5]
+   * }
+   */
+  private agentStore: Record<string, Record<Protocol, Array<NetworkAgent | undefined>>>;
+  private logger: Logger;
+
+  constructor(
+    logger: Logger,
+    private defaultConfig: HttpAgentOptions = {
+      keepAlive: true,
+      keepAliveMsecs: 50000,
+      maxSockets: 256,
+      maxFreeSockets: 256,
+      scheduling: 'lifo',
+    }
+  ) {
+    this.logger = logger;
+    this.agentStore = {};
+  }
+
+  public getAgentFactory(
+    type: string,
+    agentOptions?: HttpAgentOptions | UndiciAgentOptions | AgentFactory | false
+  ): AgentFactory {
+    if (isAgentFactory(agentOptions)) {
+      // use the user-provided factory directly
+      return agentOptions;
+    }
+
+    const agentConfig = assertValidAgentConfig(agentOptions);
+
+    // a given agent factory (of a given type) always provides the same Agent instances (for the same protocol)
+    // we keep the indexes for each protocol, so that we can access them later on, when the factory is invoked
+    const { httpInstanceIndex, httpsInstanceIndex } = this.initTypeStore(type);
+
+    return (connectionOpts: ConnectionOptions): NetworkAgent => {
+      const agents = this.getAgentInstances(type, connectionOpts);
+
+      // when the factory is called for a given connection (and protocol), we use the appropriate index
+      const agentIndex = isHttps(connectionOpts) ? httpsInstanceIndex : httpInstanceIndex;
+
+      let agent = agents[agentIndex];
+
+      if (!agent) {
+        agent = this.createAgent(agentConfig, connectionOpts);
+
+        // Allow GC'ing the Agent after it has been terminated
+        const doDestroy = agent.destroy.bind(agent);
+        agent.destroy = () => {
+          doDestroy();
+          agents[agentIndex] = undefined;
+        };
+
+        // add the new Agent instance to the list of instances of that type
+        agents[agentIndex] = agent;
+      }
+
+      return agent;
+    };
+  }
+
+  private initTypeStore(type: string): { httpInstanceIndex: number; httpsInstanceIndex: number } {
+    let agentsOfType = this.agentStore[type];
+
+    if (!agentsOfType) {
+      agentsOfType = {
+        [HTTP]: [],
+        [HTTPS]: [],
+      };
+
+      this.agentStore[type] = agentsOfType;
+    } else {
+      this.logger.warn(
+        `An Agent factory for type ${type} has already been retrieved.
+        This is not optimal since the existing instance cannot be reused.
+        Please consider reusing the previously retrieved factory instead`
```

Comment:
Why can't it be reused, again? it's because of the `destroy` call problem, or is it something else?

Reply:
The thing is, if users request an agent of a given type a second time, they are likely going to build a `new ClusterClient()` (or ES Client) with it.
If they do, they’ll have different pools with the same underlying agent.
Then, when they dispose of their `Client`, they will be closing and `destroy()`ing the underlying Agent, impacting other ES Client instances.